src/lib.rs

```diff
@@ -0,0 +1,360 @@
+                .map_err(|e| Error::ConfigError(e.to_string()))?,
+            http_client: self
+                .http_client
+                .ok_or(Error::ConfigError("missing HTTP client".to_string()))?,
+            evidence_creation_cb: self.evidence_creation_cb.ok_or(Error::ConfigError(
+                "missing evidence creation callback".to_string(),
+            ))?,
+        })
+    }
+}
+
+/// The object on which one or more challenge-response verification sessions can
+/// be run.  Always use the ChallengeResponseBuilder to instantiate it.
+pub struct ChallengeResponse {
+    base_url: url::Url,
+    http_client: reqwest::blocking::Client,
+    evidence_creation_cb: EvidenceCreationCb,
+}
+
+/// Nonce configuration: either the size (Size) of the nonce generated by the
+/// server (use 0 to let the server also pick the size of the challenge), or an
+/// explicit nonce (Value) supplied as a byte array.
+pub enum Nonce {
+    Size(usize),
+    Value(Vec<u8>),
+}
+
+impl ChallengeResponse {
+    /// Run a challenge-response verification session using the supplied nonce
+    /// configuration.  Returns the raw attestation results, or an error on
+    /// failure.
+    pub fn run(&self, nonce: Nonce) -> Result<String, Error> {
+        // create new c/r verification session on the veraison side
+        //let (session_url, session) = self.new_session_with_nonce(nonce)?;
+        let (session_url, session) = self.new_session(&nonce)?;
+
+        // invoke the user-provided evidence builder callback with per-session parameters
+        let (evidence, media_type) =
+            (self.evidence_creation_cb)(session.nonce(), session.accept())?;
+
+        // send evidence for verification to the session endpoint
+        let attestation_result = self.challenge_response(&evidence, &media_type, &session_url)?;
+
+        // return veraison's attestation results
+        Ok(attestation_result)
+    }
+
+    fn new_session(&self, nonce: &Nonce) -> Result<(String, ChallengeResponseSession), Error> {
+        // ask veraison for a new session object
+        let resp = self
+            .new_session_request(nonce)
+            .map_err(|e| Error::ApiError(e.to_string()))?;
+
+        // expect 201 and a Location header containing the URI of the newly
+        // allocated session
+        match resp.status() {
+            reqwest::StatusCode::CREATED => (),
+            status => {
+                // on error the body is a RFC7807 problem detail
+                //
+                // NOTE(tho) -- this assumption does not hold in general because
+                // the request may be intercepted (and dealt with) by HTTP
+                // middleware that is unaware of the API.  We need something
+                // more robust here that dispatches based on the Content-Type
+                // header.
+                let pd: ProblemDetails = resp.json().map_err(|e| Error::ApiError(e.to_string()))?;
```

Comment:
Am I correct in the assumption that whenever you get a `reqwest` error you want to convert it to `Error::ApiError(e.to_string())`? You could do something like
```rust
impl From<reqwest::Error> for Error {
    fn from(re: reqwest::Error) -> Self {
        Error::ApiError(re.to_string())
    }
}
```
and then you can just `?` away, without having to do the mapping "manually". Apologies for thinking about this so late, it didn't cross my mind that your errors are "organised" like this.

Reply:
And obviously a similar approach could work for any other crates with their own errors.

Reply:
> And obviously a similar approach could work for any other crates with their own errors.

true, but while *all* `reqwest::Error`s are bottled as `ApiError`s, for other error sources the mapping may be more subtle.

Reply:
done in [be6ebbb](https://github.com/veraison/rust-apiclient/pull/4/commits/be6ebbb506f46bd02b947493629f4852ec36dae0)